### PR TITLE
AO3-4852 Migrate to using RedirectExpectationHelper

### PR DESCRIPTION
**Pull request by zz9pzza**
## Issue

https://otwarchive.atlassian.net/browse/AO3-4852

## Purpose

Increase the number of checks that the tests do when running. There is a slight increase in the number of rspec tests...

## Testing

No manual testing required.

### CODE REVIEW

**Comment by sarken** [comment]
Already mentioned in chat, but wanted to leave a note here as well -- we'll merge this after we finish working on tests, when we're less likely to create merge conflicts for people who currently have work in progress.

**Comment by sarken** [comment]
Taking the deferred label off now that we've finished the big push for controller tests. We can do this whenever the merge conflicts are resolved.

File: spec/controllers/collection_participants_controller_spec.rb [hunk 1]
```diff
@@ -211,7 +215,8 @@
 
           it "displays an error notice and and redirects to collection participants" do
             put :update, params
-            it_redirects_to_with_error(collection_participants_path(collection), "Couldn't update #{participant.pseud.name}.")
+            it_redirects_to_with_error(collection_participants_path(collection),\
```
Comment (sarken) [inline spec/controllers/collection_participants_controller_spec.rb]: The `\` shouldn't be necessary here -- it's not breaking up a string

File: spec/controllers/collection_participants_controller_spec.rb [hunk 2]
```diff
@@ -430,7 +439,8 @@
           end
 
           get :add, params
-          it_redirects_to_with_error(collection_participants_path(collection), "We couldn't find anyone new by that name to add.")
+          it_redirects_to_with_error(collection_participants_path(collection),\
```
Comment (sarken) [inline spec/controllers/collection_participants_controller_spec.rb]: This shouldn't need the `\`

File: spec/controllers/gift_exchange_controller_spec.rb [hunk 3]
```diff
@@ -16,12 +17,8 @@
       expect(@collection.reload.challenge_type).to eq(nil)
     end
 
-    it "sets a flash message" do
-      expect(flash[:notice]).to eq("Challenge settings were deleted.")
-    end
-
     it "redirects to the collection's main page" do
```
Comment (sarken) [inline spec/controllers/gift_exchange_controller_spec.rb]: This should be updated to "redirects to the collection's main page with a notice" or similar now that you've combined this example and the previous one.

File: spec/controllers/works/default_rails_actions_spec.rb [hunk 4]
```diff
@@ -16,7 +16,8 @@ def call_with_params(params)
     context "when no work search parameters are given" do
       it "redirects to the login screen when no user is logged in" do
         get :clean_work_search_params, params
-        it_redirects_to new_user_session_path
+        it_redirects_to_with_error(new_user_session_path,\
```
Comment (sarken) [inline spec/controllers/works/default_rails_actions_spec.rb]: Shouldn't need a \ here or in the next change on line 184

File: spec/controllers/works/drafts_spec.rb [hunk 5]
```diff
@@ -25,36 +26,24 @@
...
+      it "should redirect to the user controller and display an appropriate error message" do
         get :drafts
-        expect(flash[:error]).to start_with "Whose drafts did you want to look at?"
+        it_redirects_to_with_error(users_path, "Whose drafts did you want to look at?")
       end
     end
 
     context "with a valid user_id" do
       context "if the user_id requested doesn't belong to the current user" do
         it "should display an error" do
```
Comment (sarken) [inline spec/controllers/works/drafts_spec.rb]: This should mention the redirect as well now that it's been combined with that example

File: spec/controllers/works/drafts_spec.rb [hunk 6]
```diff
@@ -83,30 +71,25 @@
       put :post_draft, id: random_work.id
       # There is code to return a different message in the action, but it is unreachable using a web request
       # as the application_controller redirects the user first
-      expect(flash[:error]).to start_with "Sorry, you don't have permission to access"
+      it_redirects_to_with_error(work_path(random_work),\
```
Comment (sarken) [inline spec/controllers/works/drafts_spec.rb]: Shouldn't need a \

File: spec/controllers/works/drafts_spec.rb [hunk 7]
```diff
@@ -83,30 +71,25 @@
...
         put :post_draft, id: drafts_user_work.id
-        expect(flash[:error]).to eq "That work is already posted. Do you want to edit it instead?"
-      end
-
-      it "should redirect to the user work edit page" do
-        drafts_user_work = create(:work, authors: [drafts_user.default_pseud], posted: true)
-        put :post_draft, id: drafts_user_work.id
-        expect(response).to have_http_status(:redirect)
-        expect(response).to redirect_to edit_user_work_path(drafts_user, drafts_user_work)
+        it_redirects_to_with_error(edit_user_work_path(drafts_user, drafts_user_work),\
```
Comment (sarken) [inline spec/controllers/works/drafts_spec.rb]: Shouldn't need a \, and the description of the example should be updated now that you've combined two examples

File: spec/controllers/works/drafts_spec.rb [hunk 8]
```diff
@@ -83,30 +71,25 @@
...
       end
     end
 
     it "should display an error if the work is invalid" do
       drafts_user_work = create(:work, authors: [drafts_user.default_pseud], posted: false)
       allow_any_instance_of(Work).to receive(:valid?).and_return(false)
       put :post_draft, id: drafts_user_work.id
-      expect(flash[:error]).to eq "There were problems posting your work."
-
+      it_redirects_to_with_error(edit_user_work_path(drafts_user, drafts_user_work),\
```
Comment (sarken) [inline spec/controllers/works/drafts_spec.rb]: Shouldn't need a \

File: spec/controllers/works/drafts_spec.rb [hunk 9]
```diff
@@ -117,7 +100,9 @@
       drafts_user_work.collections << draft_collection
       controller.instance_variable_set("@collection", draft_collection)
       put :post_draft, id: drafts_user_work.id
-      expect(flash[:notice]).to start_with "Work was submitted to a moderated collection."
+      it_redirects_to_with_notice(work_path(drafts_user_work),\
```
Comment (sarken) [inline spec/controllers/works/drafts_spec.rb]: Shouldn't need a \ here (but you do need it on 104)

File: spec/controllers/collection_participants_controller_spec.rb [hunk 10]
```diff
@@ -387,7 +392,8 @@
 
         it "approves those users, redirects to the collection participants page and displays a notification" do
           get :add, params
-          it_redirects_to collection_participants_path(collection)
+          expect(response).to have_http_status :redirect
```
Comment (sarken) [inline spec/controllers/collection_participants_controller_spec.rb]: Just wondering why the change here and on lines 409-410

File: spec/controllers/collection_participants_controller_spec.rb [hunk 11]
```diff
@@ -415,7 +422,9 @@
 
         it "doesn't approve the member, displays an error and redirects" do
           get :add, params
-          it_redirects_to_with_error(collection_participants_path(collection), "#{users.last.default_pseud.name} is currently banned and cannot participate in challenges.")
+          it_redirects_to_with_error(collection_participants_path(collection), \
```
Comment (sarken) [inline spec/controllers/collection_participants_controller_spec.rb]: The \ on this line shouldn't be necessary (but the next one on line 426 is necessary)

File: spec/controllers/collection_participants_controller_spec.rb (hunk 10 above)
Comment (zz9pzza) [inline spec/controllers/collection_participants_controller_spec.rb]: Looks like a bad merge I will fix it.

Comment (zz9pzza) [inline spec/controllers/collection_participants_controller_spec.rb]: It wasn't I have updated it a bit and added the following comments.
          # We use it_redirects_to_internal as it_redirects_to
          # checks that there are no extra error,notice flash messages set
          # And in this case there are.

**Comment by zz9pzza** [comment]
Do you want me to rename it_redirects_to_internal to it_redirects_to_no_checks ?

File: spec/controllers/tags_controller_spec.rb [hunk 12]
```diff
@@ -183,7 +180,8 @@
         put :update, id: tag, tag: { canonical: false }, commit: "Wrangle"
         tag.reload
         expect(tag.canonical?).to be_falsy
-        it_redirects_to wrangle_tag_path(tag, page: 1, sort_column: "name", sort_direction: "ASC")
+        it_redirects_to_with_notice wrangle_tag_path(tag, page: 1, sort_column: "name", sort_direction: "ASC"),
```
Comment (sarken) [inline spec/controllers/tags_controller_spec.rb]: Can this be written in a style that's consistent with all the other tests? That is: 
```
it_redirects_to_with_notice(wrangle_tag_path(tag, page: 1, sort_column: "name", sort_direction: "ASC"),
                                    "Tag was updated.")
```

File: spec/controllers/locales_controller_spec.rb [hunk 13]
```diff
@@ -50,7 +50,8 @@
 
       it "redirects to the user page" do
         put :update
-        it_redirects_to user_path(user)
+        it_redirects_to_with_error user_path(user),
```
Comment (sarken) [inline spec/controllers/locales_controller_spec.rb]: Could these be rewritten with parentheses to be consistent with the other examples? Also, the descriptions should mention the flash errors.

File: spec/controllers/inbox_controller_spec.rb [hunk 14]
```diff
@@ -8,13 +8,15 @@
   describe "GET #show" do
     it "redirects to user page when not logged in" do
       get :show, user_id: user.login
-      it_redirects_to user_path(user)
```
Comment (sarken) [inline spec/controllers/inbox_controller_spec.rb]: Could these be rewritten with parentheses to be consistent with the majority of the other examples? Also, the descriptions should be updated to mention the flash messages.

File: spec/controllers/collection_participants_controller_spec.rb [hunk 15]
```diff
@@ -400,7 +406,10 @@
       context "where the users to be added haven't yet applied to the collection" do
         it "creates new participants with the member role and redirects" do
           get :add, params
-          it_redirects_to collection_participants_path(collection)
+          # We use it_redirects_to_simple as it_redirects_to
+          # checks that there are no extra error,notice flash messages set
```
Comment (sarken) [inline spec/controllers/collection_participants_controller_spec.rb]: There should be a space after the comma here. (Also, it might just be me, but I think "because" might be clearer than "as" here -- I initially thought this meant `it_redirects_to_simple` was acting as an alias for `it_redirects_to`.)

File: spec/controllers/admin/api_controller_spec.rb [hunk 16]
```diff
@@ -24,7 +24,7 @@
 
       it "redirects to the homepage" do
         get :index, params
-        it_redirects_to root_path
+        it_redirects_to_with_notice(root_path,"I'm sorry, only an admin can look at that area")
```
Comment (sarken) [inline spec/controllers/admin/api_controller_spec.rb]: In addition to Hound's comment about the comma, the descriptions for these two examples should ideally mention the flash notice.

**Comment by sarken** [comment]
Re: "Do you want me to rename it_redirects_to_internal to it_redirects_to_no_checks ?"

It looks like you named it it_redirects_to_simple, which works for me!

File: spec/controllers/collection_participants_controller_spec.rb (hunk 10 above)
Comment (redsummernight) [inline spec/controllers/collection_participants_controller_spec.rb]: Now it's 2 `expect`s instead of `it_redirects_to_simple`.

Comment (zz9pzza) [inline spec/controllers/collection_participants_controller_spec.rb]: That is because we are not certain how the users will be added to the notice method.

**Comment by zz9pzza** [comment]
I have updated the descriptions in  spec/controllers/inbox_controller_spec.rb which are the ones I could see that were outstanding and commented on @redsummernight question ?

**Comment by sarken** [comment]
Alas, this has merge conflicts now.

**Comment by sarken** [comment]
Now at #3429
